Resolve Sorting MultiValues Problem HSEARCH-3839

wklaczynski opened the pull request.

yrodiere commented on backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/sort/impl/ElasticsearchDistanceSortBuilder.java:
```diff
@@ -39,4 +40,9 @@ protected void doBuildAndAddTo(ElasticsearchSearchSortCollector collector, JsonO
+
+	@Override
+	public void multi(SortMultiFunc multi) {
+		throw new UnsupportedOperationException( "Not supported yet." ); //To change body of generated methods, choose Tools | Templates.
```
This looks incomplete?

yrodiere commented on backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/sort/impl/ElasticsearchFieldSortBuilder.java:
```diff
@@ -108,4 +109,9 @@ public void doBuildAndAddTo(ElasticsearchSearchSortCollector collector, JsonObje
+
+	@Override
+	public void multi(SortMultiFunc multi) {
+		throw new UnsupportedOperationException( "Not supported yet." ); //To change body of generated methods, choose Tools | Templates.
```
This looks incomplete?

yrodiere commented on backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/sort/impl/ElasticsearchScoreSortBuilder.java:
```diff
@@ -28,4 +29,9 @@ public void doBuildAndAddTo(ElasticsearchSearchSortCollector collector, JsonObje
+
+	@Override
+	public void multi(SortMultiFunc multi) {
+		throw new UnsupportedOperationException( "Not supported yet." ); //To change body of generated methods, choose Tools | Templates.
```
This looks incomplete?

wklaczynski replied:
I suspect there is a solution. The more that I used the code from elasticsearch. It is difficult for me to solve the problem here because I do not use elasticsearch. In such cases, I hope that the right person will complete this code. The exception gives this information.

wklaczynski replied:
Changed, I set the parameter in the building material. The multi value is set, similarly in the numeric range and terms aggregation. In Apache Lucene this has been resolved, but in Elasticsearch it has no effect. But it can be used if there is a similar option in Elasticsearch.

yrodiere commented on backend/elasticsearch/nb-configuration.xml:
```diff
@@ -0,0 +1,35 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Please don't include IDE configuration in your commits. You can add this line to the `.gitignore` file to have these files automatically ignored:

```
nb-configuration.xml
```

wklaczynski replied:
That would require a global patch in .gitignore.

`# This patch file was generated by NetBeans IDE
# It uses platform neutral UTF-8 encoding and \n newlines.
--- a/.gitignore
+++ b/.gitignore
@@ -45,3 +45,7 @@
 # Maven wrapper files
 .mvn
 mvnw*
+/backend/elasticsearch/nb-configuration.xml
+/backend/lucene/nb-configuration.xml
+/engine/nb-configuration.xml
+/nb-configuration.xml
\ No newline at end of file`

yrodiere replied:
Alright, I added it to `.gitignore`. Please remove the files from your commit and rebase on master: they should be ignored by git from now on.

yrodiere commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/lowlevel/docvalues/impl/AbstractNumericDocValues.java:
```diff
@@ -0,0 +1,37 @@
+ * Base implementation that throws an {@link IOException} for the
+ * {@link DocIdSetIterator} APIs. This impl is safe to use for sorting and
+ * aggregations, which only use {@link #advanceExact(int)} and
+ * {@link #longValue()}.
```
Since this was copied from Elasticsearch (I think?), please give them credit, to avoid legal trouble:

```
 * <p>
 * Copied and adapted from {@code PUT THE ORIGINAL CLASS NAME HERE}
 * of <a href="https://github.com/elastic/elasticsearch">Elasticsearch project</a>.
```

wklaczynski replied:
From what I see, a lot has been copied from there, practically the whole package is a copy. I corrected in the package where there was no such information.

wklaczynski replied:
Corrected

yrodiere commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/lowlevel/docvalues/impl/AbstractSortedDocValues.java:
```diff
@@ -0,0 +1,35 @@
+ * Base implementation that throws an {@link IOException} for the
+ * {@link DocIdSetIterator} APIs. This impl is safe to use for sorting and
+ * aggregations, which only use {@link #advanceExact(int)} and
+ * {@link #ordValue()}.
```
Same here, please add a few lines to give credit.

wklaczynski replied:
It was added earlier.

yrodiere commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/lowlevel/docvalues/impl/AbstractSortedNumericDocValues.java:
```diff
@@ -0,0 +1,40 @@
+ * Base implementation that throws an {@link IOException} for the
+ * {@link DocIdSetIterator} APIs. This impl is safe to use for sorting and
+ * aggregations, which only use {@link #advanceExact(int)} and
+ * {@link #docValueCount()} and {@link #nextValue()}.
```
Same here, please add a few lines to give credit.

wklaczynski replied:
Corrected

yrodiere commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/lowlevel/docvalues/impl/FieldData.java:
```diff
@@ -0,0 +1,339 @@
+import org.apache.lucene.index.SortedNumericDocValues;
+
+/**
+ * Utility methods, similar to Lucene's {@link DocValues}.
```
Same here, please add a few lines to give credit.

wklaczynski replied:
Corrected

yrodiere commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/lowlevel/docvalues/impl/MultiValueMode.java:
```diff
@@ -0,0 +1,757 @@
+
+/**
+ * Defines what values to pick in the case a document contains multiple values
+ * for a particular field.
```
Same here, please add a few lines to give credit.

wklaczynski replied:
Corrected

yrodiere commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/codec/impl/AbstractLuceneNumericFieldCodec.java:
```diff
@@ -85,15 +85,15 @@ public boolean isCompatibleWith(LuceneFieldCodec<?> obj) {
-		return ( projectable == other.projectable ) && ( searchable == other.searchable )
-				&& ( sortable == other.sortable ) && ( aggregable == other.aggregable );
+		return (projectable == other.projectable) && (searchable == other.searchable)
+			&& (sortable == other.sortable) && (aggregable == other.aggregable);
```
This probably won't pass the style checks.

wklaczynski replied:
I compiled with control. I always have style control turned on. Passed correctly. But it always adjusts the autoformat parameters. Unfortunately, there is no setting for this style.

yrodiere commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/lowlevel/impl/LuceneDoubleDomain.java:
```diff
@@ -104,21 +111,56 @@ public IndexableField createDocValuesField(String absoluteFieldPath, Double nume
 		}
 	}
+
+	public class SortedDoubleDocValuesField extends SortedNumericDocValuesField {
```
I think this class should be `static`?

wklaczynski replied:
They really should be static. Correct.

wklaczynski replied:
It's supposed to be static. Corrected.

yrodiere commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/lowlevel/impl/LuceneFloatDomain.java:
```diff
@@ -105,21 +112,56 @@ public IndexableField createDocValuesField(String absoluteFieldPath, Float numer
 		}
 	}
+
+	public class SortedFloatDocValuesField extends SortedNumericDocValuesField {
```
I think this class should be `static`?

wklaczynski replied:
It's supposed to be static. Corrected.

wklaczynski replied:
Corrected

yrodiere commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/sort/impl/AbstractLuceneStandardFieldSortBuilder.java:
```diff
@@ -49,11 +50,11 @@
+		String absoluteFieldPath, String nestedDocumentPath,
+		DslConverter<?, ? extends F> converter, DslConverter<F, ? extends F> rawConverter,
+		LuceneCompatibilityChecker converterChecker, C codec,
+		Object sortMissingValueFirstPlaceholder, Object sortMissingValueLastPlaceholder) {
```
While I can imagine why you prefer to indent code this way, this does not fit the style guide of this project. Please don't reindent code like this.

You can find information on how to set up your IDE on our website: [here for Eclipse](https://hibernate.org/community/contribute/eclipse-ide/) and [here for IDEA](https://hibernate.org/community/contribute/intellij-idea/). I do not know about netbeans, though (that's what you are using, right?)

wklaczynski replied:
Little Netbeans is in some cases automated. It helps a lot while programming, but unfortunately it doesn't give 100% compatibility with Eclipse. It's hard to see what a formatter will do, I rely on "maven-checkstyle-plugin". The differences are not intentional. Result from auto text formatting.

wklaczynski replied:
I found a parameter. Corrected.

wklaczynski replied:
Corrected

yrodiere commented on engine/src/main/java/org/hibernate/search/engine/search/sort/spi/SearchMultivalueBuilder.java:
```diff
@@ -0,0 +1,28 @@
+ * @param <B> The implementation type of the builder, which should expose a {@code build()} method.
+ * This type is backend-specific, as the parameters to the build method may vary from one backend to another.
+ */
+public interface SearchMultivalueBuilder<B> {
```
I think this interface is not used an can be removed?

wklaczynski replied:
Not really needed. Copy from SearchSortBuilder. I will remove.

wklaczynski commented on integrationtest/backend/lucene/src/test/java/org/hibernate/search/integrationtest/backend/lucene/search/LuceneMultiValueSearchPredicateIT1.java:
```diff
@@ -0,0 +1,927 @@
+			.toQuery();
+
+		assertThat( query ).hasDocRefHitsExactOrder( c -> {
+			c.doc( INDEX_NAME_1_1, DOCUMENT_1_1_2, DOCUMENT_1_1_1 );
```
I know the data sample is too small. But whatever happens in the code there will be an exception rather than a test done differently. The test checks the behavior inside, tests all possible modes of work inside the code. There are currently four types of variables (Double, Float, Long, Integer). Including three possible ways to perform these functions (Master, Flattended, Nested). In each of these cases, different code is executed.

wklaczynski commented:
**Regarding the filtering of sort values according to the predicate**

Each user cannot see more than he is authorized to. It looks weird now. If even one price is available then the document is found, it displays the correct price, but not in its place in order. Of course, my application is careful not to display the price unavailable to the user, but the sorting is not correct. Of course, I could sort it again programmatically, but I would have to download all results and re-sort them, choose only as much as I can display on the page, it would be inefficient, so probably database indexes would be faster.

For me it is a matter if hibernate-search I will be able to use it in my applications. I really like the simplification of the code provided by hibernate-search, it takes a long time to do the same in apache lucene. Great work, what you do is a lot of good work. But it must implement just such complex problems to be useful. I think it can be solved, the argent method a pache JoinUtil.createJoinQuery does just that, only relatively on two indexes. but it should also work the same way on one index. I will try to solve it, I think it will not take much time. Only to get to this point I had to fix aggregation and sorting. As for sorting, there was an attempt to deal with many values, there was no attempt at aggregation to solve this problem. Currently, supper sorting and aggregation work for me. Now I'll look at "NestedDocsProvider", why it returns all nested documents. As for the operating modes, you can always give a switch, who needs what they will set in the query.

wklaczynski commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/lowlevel/docvalues/impl/FieldData.java:
```diff
@@ -13,7 +13,10 @@
  * Utility methods, similar to Lucene's {@link DocValues}.
- */
+ * <p>
+ * Copied and adapted from {@code PUT THE ORIGINAL CLASS NAME HERE}
```
I have removed "FieldData.java". I use a different solution. Everything is "DoubleMultiValuesSource" and "LongMultiValuesSource". This can sort and aggregate.

wklaczynski commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/lowlevel/docvalues/impl/AbstractSortedNumericDocValues.java:
```diff
@@ -14,6 +14,9 @@
  * aggregations, which only use {@link #advanceExact(int)} and
  * {@link #docValueCount()} and {@link #nextValue()}.
+ * <p>
+ * Copied and adapted from {@code PUT THE ORIGINAL CLASS NAME HERE}
```
I have removed "AbstractSortedNumericDocValues.java". I use a different solution. Everything is "DoubleMultiValuesSource" and "LongMultiValuesSource". This can sort and aggregate.

wklaczynski commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/lowlevel/docvalues/impl/AbstractSortedDocValues.java:
```diff
@@ -14,7 +14,10 @@
  * {@link #ordValue()}.
- */
+ * <p>
+ * Copied and adapted from {@code PUT THE ORIGINAL CLASS NAME HERE}
```
I have removed "AbstractSortedDocValues.java ". I use a different solution. Everything is "DoubleMultiValuesSource" and "LongMultiValuesSource". This can sort and aggregate.

wklaczynski commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/lowlevel/docvalues/impl/AbstractNumericDocValues.java:
```diff
@@ -16,6 +16,9 @@
  * aggregations, which only use {@link #advanceExact(int)} and
  * {@link #longValue()}.
+ * <p>
+ * Copied and adapted from {@code PUT THE ORIGINAL CLASS NAME HERE}
```
I have removed "AbstractNumericDocValues.java ". I use a different solution. Everything is "DoubleMultiValuesSource" and "LongMultiValuesSource". This can sort and aggregate.

wklaczynski commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/lowlevel/docvalues/impl/MultiValueMode.java:
```diff
@@ -20,7 +20,10 @@
  * for a particular field.
- */
+ * <p>
+ * Copied and adapted from {@code PUT THE ORIGINAL CLASS NAME HERE}
```
Removed completely.

wklaczynski commented:
I suggest you do everything as in Elesticsearch. I'm going to replace the "multi" predicate with "mode". For sort and aggregate, I will add the option to add a filter, which will create a filter for the nested query in the nested provider. The whole solution can be easily added to elestcsearch.

https://discuss.elastic.co/t/sorting-by-nested-field/103141

yrodiere commented:
> For sort and aggregate, I will add the option to add a filter, which will create a filter for the nested query in the nested provider

Looks good to me. This will require deep changes to the DSL, in order to provide the ability to define  predicates within the sort DSL. Which means a lot of new changes on top of those already in this PR. Please, could you consider finishing the work on multi-values without filter first (at least having it build and having pre-existing tests pass for Lucene), so that we can review this already considerable amount of code and merge it?

You're fast, and this is great, but review/polishing work is piling up, and the more there is, the harder it will be to catch up once you're done.

wklaczynski commented:
Sorting and aggregation of many values is completed. I will open a new branch for filtering values before sorting. The last thing I changed was the parameter name in the dsl predicate and I added sort implementations to Elasticsearch.

yrodiere commented:
Ok. I'll try to organize the commits a bit and see if I can make the tests pass. Closing this one, I'll reopen another (authoring information will be preserved for your commits, of course).

@wklaczynski For your next PR, could you please follow the recommendations in [CONTRIBUTING.md](https://github.com/hibernate/hibernate-search/blob/master/CONTRIBUTING.md#commit)? It'll make merging your PRs that much easier. In particular:

* Organize the commits in logical units. For example "Add API for feature X", "Implement feature X for Lucene", "Test feature X", "Document feature X".
* Format your commit messages using the format "HSEARCH-XXXX <message>"

Thanks in advance.
